### ISSUE TITLE
online deduplication

### ISSUE THREAD

**Issue by vtolstov** [sheepdog/sheepdog#140]
online deduplication can bring some good things like:
- minimize storage workload
- minimize disk usage
- speedup vm writing

@mitake what  needs to be added to support this feature?

**Comment by mitake** [comment]
It is hard to implement dedup because it changes relation between inodes and objects. There are bunch of things if you want to implement it (I cannot list up soon). In addition, effect of dedup heavily depends on usage of sheepdog.

How about implement dry run version of dedup as a first step? e.g. dog vdi dedup \<root VDI name\> If the command can calculate amount of reduced disk space if we execute dedup on the VDI family (I suggest limiting the scope of dedup to VDI family unit for simplify), it will be useful because we can know the effect of dedup.

**Comment by marcin-github** [comment]
zfs has deduplication, it needs a lot of ram to track hash of every written block. If sheepdog would have online deduplication then requirements of RAM for sheepdog will be much, much higher. Maybe it would be better to have offline dedup?

**Comment by vtolstov** [comment]
But does it possible to before write some block calculate it checksum and try to find this block via calculated checksum?

**Comment by marcin-github** [comment]
Yes but table contains hashes of blocks can be big, where it should be kept, on disk (slow performance), in memory (on every node, on one node, partitioned beetwen nodes? how much memory sheep would needs?). Please look at https://wiki.freebsd.org/ZFSTuningGuide#Deduplication , I'm not sure if sheepdog can handle dedup in the less-resources-needed way .

**Comment by sirio81** [comment]
I also think on-line dedup is going to be to expensive in terms of resources.
An offline dedup may be interesting.
But if this feature may have impact in stability or risk to corrupt
vdi, I would wait to implement it.
There are already many features that requires more testing and fixing.

**Comment by vtolstov** [comment]
As i see , yes, for online dedup we need gateway nodes with big memory, that not always possible, so offline dedup may be preferred.

**Comment by mitake** [comment]
Yes, offline dedup (or its dry run mode) would be a good start point if we really work on it.
